### DISCUSSION TITLE
Consider removing put() method in AuthenticationKeyManager

### DISCUSSION THREAD

**Discussion by lukasbalaz** [serverpod/serverpod#2811] [Q&A] [answered]
I am new to serverpod and at the moment i am implementing my own authentication (loosely following https://docs.serverpod.dev/concepts/authentication/custom-overrides).

I've already got it working, I have just one question / suggestion. The way I structured my client code, the api key is always coupled with userId. It doesn't make sense to be logged in without knowing the userId.

When implementing my own `AuthenticationKeyManager`, I've got no problem with overriding `get()` and `remove()` methods but I am not sure what to do with the `put()` method - I can't just change the stored apiKey without changing the userId, also.

I also noticed that the method is not used in the code (at least VSCode didn't find any references in `serverpod_client` package). What is the intended use of this method so that I can implement it properly? And if there is none, couldn't it just be removed?

EDIT: Link to the code:
https://github.com/serverpod/serverpod/blob/main/packages/serverpod_client/lib/src/auth_key_manager.dart#L9

**Comment by SandPod** [comment, marked as answer]
Hey @lukasbalaz! 👋 

Happy to hear you are exploring Serverpod!

The purpose of the `put()` method is to store an authentication key in the `AuthenticationKeyManager`. It is most often implemented by the code managing the login flow.

In the Serverpod authentication module, you can find a call to `put()`in the `SessionManager`´s (Responsible for managing the users state) `registerSignedInUser` method, this method is then called in all login flows for the different authentication providers (E.g. Apple, email, google, etc...) : https://github.com/serverpod/serverpod/blob/d93a323f150601b5933603bb05ee74fbf416e40a/modules/serverpod_auth/serverpod_auth_shared_flutter/lib/src/session_manager.dart#L56-L65
